File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int buyorsell(int idx, boolean buy, int[] prices, int[][] dp){
        if(idx == prices.length){
            return 0;
        }
        
        if(dp[idx][buy ? 1:0] != -1) return dp[idx][buy ? 1:0];

        int profit = 0;
        if(buy){
            profit = Math.max(-prices[idx] + buyorsell(idx+1, false, prices, dp),
                                    buyorsell(idx+1, true, prices, dp));
        }
        else{
            profit = Math.max(prices[idx] + buyorsell(idx+1, true, prices, dp),
                                    buyorsell(idx+1, false, prices, dp));
        }

        return dp[idx][buy ? 1:0] = profit;
    }

    public int maxProfit(int[] prices) {
        int[][] dp = new int[prices.length+1][2];
        dp[prices.length][1] = 0;
        dp[prices.length][1] = 0;

        for(int i=prices.length-1; i>=0; i--){
            int profit;
            for(int buy =0; buy<=1; buy++){
                if(buy == 1){
                    profit = Math.max(-prices[i] + dp[i+1][0], dp[i+1][1]);
                }
                else{
                    profit = Math.max(prices[i] + dp[i+1][1],
                                    dp[i+1][0]);
                }
                dp[i][buy] = profit;
            }
        }

        return dp[0][1];

    }
}

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)

        # def dfs(idx, profit, haveStock):
        #     if idx == n:
        #         return profit

        #     #Buy
        #     buy = sell = 0
        #     if haveStock == False:
        #         buy = max(dfs(idx+1, profit-prices[idx], True), dfs(idx+1, profit, False))

        #     #Sell
        #     else:
        #         sell = max(dfs(idx+1, profit+prices[idx], False), dfs(idx+1, profit, True))

        #     return max(buy, sell)

        dp = [[-1] * 2] * (n+1)
        dp[n][0] = 0
        dp[n][1] = 0

        for i in range(n-1, -1, -1):
            #Sell
            profit = max(prices[i]+dp[i+1][1], dp[i+1][0])
            dp[i][0] = profit

            #Buy
            profit = max(-prices[i]+dp[i+1][0], dp[i+1][1])
            dp[i][1] = profit

        return dp[0][1]

